File: frontend/apps/cloudity-web/src/pages/app/mail/mailViewPreferences.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { loadMailViewState, saveMailViewState } from './mailViewPreferences'

describe('mailViewPreferences', () => {
  beforeEach(() => {
    localStorage.clear()
  })

  it('persiste boîte et dossier par tenant + email', () => {
    saveMailViewState(1, 'User@Test.com', { accountId: 42, folder: 'unified' })
    expect(loadMailViewState(1, 'user@test.com')).toEqual({ accountId: 42, folder: 'unified' })
    expect(loadMailViewState(2, 'user@test.com')).toEqual({ accountId: null, folder: 'inbox' })
  })

  it('migre l’ancienne clé account-only', () => {
    localStorage.setItem('cloudity_mail_selected_account_id', '7')
    expect(loadMailViewState(1, 'a@test.com')).toEqual({ accountId: 7, folder: 'inbox' })
  })
})

File: mobile/contacts/test/mail_view_preferences_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'package:cloudity_shared/mail_view_preferences.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() async {
    SharedPreferences.setMockInitialValues({});
  });

  test('persiste boîte et dossier par tenant + email', () async {
    await MailViewPreferences.save(
      email: 'User@Test.com',
      tenantId: 2,
      accountId: 42,
      folder: 'sent',
    );
    final loaded = await MailViewPreferences.load(
      email: 'user@test.com',
      tenantId: 2,
    );
    expect(loaded.accountId, 42);
    expect(loaded.folder, 'sent');
  });

  test('migre l’ancienne clé mobile (email seul)', () async {
    SharedPreferences.setMockInitialValues({
      'cloudity.mail.view.v1:user@test.com':
          '{"accountId":7,"folder":"scheduled"}',
    });
    final loaded = await MailViewPreferences.load(
      email: 'user@test.com',
      tenantId: 1,
    );
    expect(loaded.accountId, 7);
    expect(loaded.folder, 'scheduled');
  });

  test('dossier invalide retombe sur inbox', () async {
    await MailViewPreferences.save(
      email: 'a@test.com',
      tenantId: 1,
      accountId: 1,
      folder: 'x' * 600,
    );
    final loaded = await MailViewPreferences.load(
      email: 'a@test.com',
      tenantId: 1,
    );
    expect(loaded.folder, 'inbox');
  });
}
